Cache.cfc

```diff
@@ -0,0 +1,367 @@
+				// our cache is empty - set an empty resultsArray
+				local.resultsArray = [];
+			} else {
+				rethrow;
+			}
+		} catch(org.terracotta.toolkit.nonstop.NonStopException e) {
+			// terracotta-backed caches throw this exception during a service interruption
+			local.cacheException = true;
+			local.resultsArray = [];
+		}
+
+		// queryNew only supports a subset of the data types that queryparam does
+		local.fieldSQLTypes = listReduce(
+			arguments.selectStatement.getSelect(),
+			function(v, i) {
+				switch(getFieldSQLType(i)) {
+					case "char":
+						return listAppend(v, "varchar");
+						break;
+					case "float":
+					case "money":
+					case "numeric":
+					case "real":
+						return listAppend(v, "double");
+						break;
+					case "smallint":
+					case "tinyint":
+						return listAppend(v, "integer");
+						break;
+					default:
+						return listAppend(v, getFieldSQLType(i));
+						break;
+				};
+			},
+			""
+		);
+
+		local.query = queryNew(arguments.selectStatement.getSelect(), local.fieldSQLTypes);
+
+		for(local.result in local.resultsArray) {
+			local.row = {};
+
+			for(local.column in arguments.selectStatement.getSelect()) {
+				local.value = local.result.getAttribute(getInstance().getSearchAttribute(local.column));
+
+				if(structKeyExists(local, "value")) {
+					switch(getFieldSQLType(local.column)) {
+						case "date":
+						case "time":
+						case "timestamp":
+							// date/time values need to be explicitly manipulated to get the precision we need
+							if(local.value >= 0) {
+								local.row[local.column] = createObject("java", "java.util.Date").init(local.value);
+							} else {
+								local.row[local.column] = javaCast("null", "");
+							}
+							break;
+						default:
+							// the rest of our data types should be coerced correctly based on the column type defined above
+							local.row[local.column] = local.value;
+							break;
+					};
+				} else {
+					local.row[local.column] = javaCast("null", "");
+				}
+			}
+
+			queryAddRow(local.query, local.row);
+		}
+
+		local.query
+			.getMetadata()
+				.setExtendedMetadata({
+					cached: !structKeyExists(local, "cacheException"),
+					recordCount: arrayLen(local.resultsArray),
+					totalRecordCount: (structKeyExists(local, "results") ? local.results.size() : 0)
+				});
+
+		// clean up after ourselves
+		if(structKeyExists(local, "results")) {
+			local.results.discard();
+		}
+
+		return local.query;
+	}
+
+	boolean function fieldExists(required string fieldName) {
+		queryableCheck();
+
+		return variables.queryable.fieldExists(arguments.fieldName) && variables.queryable.fieldIsFilterable(arguments.fieldName);
+	}
+
+	boolean function fieldIsFilterable(required string fieldName) {
+		queryableCheck();
+
+		return variables.queryable.fieldIsFilterable(arguments.fieldName);
+	}
+
+	string function getFieldList() {
+		queryableCheck();
+
+		return listReduce(
+			variables.queryable.getFieldList(),
+			function(v, i) {
```

Comment:
@timmixell Not sure if we want more descriptive variable names here or not

Reply:
i'm gonna draw the line here - i would compare this to declaring for-loop vars